src/main/java/org/testng/internal/annotations/IgnoreListener.java

```diff
@@ -0,0 +1,71 @@
+package org.testng.internal.annotations;
+
+import org.testng.annotations.ITestAnnotation;
+import org.testng.annotations.Ignore;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+
+public class IgnoreListener implements IAnnotationTransformer {
+
+    @Override
+    public void transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) {
+        if (!annotation.getEnabled()) {
+            return;
+        }
+        Class<?> typedTestClass = testClass;
+        if (testMethod != null) {
+            ignoreTest(annotation, testMethod.getAnnotation(Ignore.class));
+            typedTestClass = testMethod.getDeclaringClass();
+        }
+        if (typedTestClass != null) {
+            ignoreTest(annotation, typedTestClass.getAnnotation(Ignore.class));
+            Package testPackage = typedTestClass.getPackage();
+            if (testPackage != null) {
+                ignoreTest(annotation, findAnnotation(testPackage));
+            }
+        }
+    }
+
+    private static void ignoreTest(ITestAnnotation annotation, Ignore ignore) {
```

Comment:
This is not a review comment, but more of a question. 

Do you think we can maybe write this method to not include nesting of conditions ? [ No, there's nothing wrong in that, but I have seen that nesting of conditions and loops over a period of time starts to add complexity to the method in terms of understandability ]. 

Of-course my version has multiple returns and I also agree that this is more of like a programming style. But still wanted to know what your thoughts on this were.

```java
private static void ignoreTest(ITestAnnotation annotation, Ignore ignore) {
    if (ignore == null) {
        return;
    }
    annotation.setEnabled(false);
    if (ignore.value().isEmpty()) {
        return;
    }
    String ignoredDescription;
    if (annotation.getDescription() == null || annotation.getDescription().isEmpty()) {
        ignoredDescription = ignore.value();
    } else {
        ignoredDescription = ignore.value() + ": " + annotation.getDescription();
    }
    annotation.setDescription(ignoredDescription);    
}
```

Reply:
I prefer your version too 👍 
I will update.

Reply:
done